sql/src/main/java/io/crate/operation/reference/sys/check/cluster/TablesNeedUpgradeSysCheck.java

```diff
@@ -0,0 +1,143 @@
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.operation.reference.sys.check.cluster;
+
+import io.crate.action.sql.Option;
+import io.crate.action.sql.ResultReceiver;
+import io.crate.action.sql.SQLOperations;
+import io.crate.data.Row;
+import io.crate.operation.reference.sys.check.AbstractSysCheck;
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.cluster.ClusterService;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.inject.Singleton;
+import org.elasticsearch.common.logging.ESLogger;
+import org.elasticsearch.common.logging.Loggers;
+
+import javax.annotation.Nonnull;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.concurrent.CompletableFuture;
+
+@Singleton
+public class TablesNeedUpgradeSysCheck extends AbstractSysCheck {
+
+    public static final int ID = 4;
+    public static final String DESCRIPTION =
+        "The following tables must be upgraded for compatibility with future versions of CrateDB: ";
+
+    private static final ESLogger LOGGER = Loggers.getLogger(TablesNeedUpgradeSysCheck.class);
+    private static final String STMT = "select schema_name || '.' || table_name, min_lucene_version " +
+                                       "from sys.shards where min_lucene_version <> '" +
+                                       org.apache.lucene.util.Version.LATEST.toString() + "' " +
+                                       "AND (? = [] OR schema_name || '.' || table_name != ANY (?)) " +
+                                       "order by 1";
+    private static final int LIMIT = 50_000;
+    private static final String PREP_STMT_NAME = "tables_need_recrate_syscheck";
+
+    private final SQLOperations.Session session;
+    private final ClusterService clusterService;
+    private volatile Collection<String> tablesNeedUpgrade;
+
+    @Inject
+    public TablesNeedUpgradeSysCheck(ClusterService clusterService, SQLOperations sqlOperations) {
+        super(ID, DESCRIPTION, Severity.MEDIUM);
+        this.clusterService = clusterService;
+        this.session = sqlOperations.createSession("sys", Option.NONE, LIMIT);
+        this.session.parse(PREP_STMT_NAME, STMT, Collections.emptyList());
+    }
+
+    @Override
+    public BytesRef description() {
+        String linkedDescriptionBuilder = DESCRIPTION + tablesNeedUpgrade + ' ' + LINK_PATTERN + ID;
+        return new BytesRef(linkedDescriptionBuilder);
+    }
+
+    @Override
+    public CompletableFuture<?> computeResult() {
+        Collection<String> tablesNeedRecreation =
+            LuceneVersionChecks.tablesNeedRecreation(clusterService.state().metaData());
+
+        final CompletableFuture<Collection<String>> result = new CompletableFuture<>();
+        try {
+            Object[] sqlParam = tablesNeedRecreation.toArray(new String[]{});
```

Comment:
would it make sense to somehow share the impl with AlterTableOperation.executeAlterTableAddColumn